### PR TITLE
Merge implementation classes with interfaces in Python

### PR DESCRIPTION
**Pull request by pepone**
This PR merges the implementation and abstract classes for Communicator, ObjectAdpater, Properties, and ImplicitContext.

This also includes the `Ice.Current` changes from  #2552, and the doc formatting changes for the affected classes from #2535

### CODE REVIEW

File: python/python/Ice/Communicator.py [hunk 1]
```diff
@@ -1,24 +1,53 @@
...
 
-__name__ = "Ice"
+from .ObjectAdapter import ObjectAdapter
+from .ImplicitContext import ImplicitContext
+from .Properties import Properties
+from .LoggerI import LoggerI
+from .Logger import Logger
 
-class Communicator(object):
+class Communicator:
```
Comment (bernardnormier) [inline python/python/Ice/Communicator.py:9]: Should we mark it as `@final` like in C++, C# etc.?

File: python/python/Ice/Communicator.py [hunk 2]
```diff
@@ -1,24 +1,53 @@
...
+        return self._impl
 
     def destroy(self):
         """
         Destroy the communicator. This operation calls shutdown implicitly. Calling destroy cleans up
         memory, and shuts down this communicator's client functionality and destroys all object adapters. Subsequent
         calls to destroy are ignored.
         """
-        raise NotImplementedError("method 'destroy' not implemented")
+        self._impl.destroy()
```
Comment (bernardnormier) [inline python/python/Ice/Communicator.py:50]: (Not from this PR). I find the description misleading. We should never use "shut down the communicator's client functionality". Shut down is only about server-side functionality.

**Review by bernardnormier — APPROVED** [review]
Looks good.

File: python/python/Ice/Communicator.py (hunk 1 above)
Comment (pepone) [reply python/python/Ice/Communicator.py:11]: Fixed. I added the `@final` decorator.

**Pull request merged**